### PR TITLE
MAILET-135 Add a simple test on sending an email

### PR DESCRIPTION
**Pull request by aduprat**
In complement of #405
Only 4 last commits, the first one `MAILET-135 Clone mail when adding to MemoryMailQueue` was a bug in InMemory implementation

### CODE REVIEW

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java [hunk 1]
```diff
@@ -78,12 +79,18 @@ public void enQueue(Mail mail, long delay, TimeUnit unit) throws MailQueueExcept
...
-                mailItems.addFirst(new MemoryMailQueueItem(new MailImpl(mail)));
+                mailItems.addFirst(new MemoryMailQueueItem(cloneMail(mail)));
             } catch (MessagingException e) {
                 throw new MailQueueException("Error while copying mail " + mail.getName(), e);
             }
         }
 
+        private Mail cloneMail(Mail mail) throws MessagingException {
+            MailImpl mailImpl = new MailImpl(mail);
+            mailImpl.setMessage(new MimeMessage(mail.getMessage()));
```
Comment (mbaechler) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: this is EVIL :smiling_imp:

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java [hunk 2]
```diff
@@ -85,6 +87,7 @@ public void process(Exchange exchange) throws Exception {
             }
 
         } finally {
+            MailetPipelineLogging.logOut(mailet.getMailetConfig().getMailetContext().getLogger(), mailet.getMailetInfo());
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java]: what about logging mail state before and after ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 3]
```diff
@@ -49,15 +67,78 @@ public void setup() throws Exception {
...
+    @Test
+    public void simpleMailShouldBeSent() throws Exception {
+        jamesServer.getServerProbe().addDomain("james.org");
+        String from = "user@james.org";
+        jamesServer.getServerProbe().addUser(from, PASSWORD);
+        String recipient = "user2@james.org";
+        jamesServer.getServerProbe().addUser(recipient, PASSWORD);
+        jamesServer.getServerProbe().createMailbox(MailboxConstants.USER_NAMESPACE, recipient, "INBOX");
+
+        try (SocketChannel socketChannel = SocketChannel.open();) {
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: you can remove the `;`

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 4]
```diff
@@ -49,15 +67,78 @@ public void setup() throws Exception {
...
+                    ".\r\n";
+            socketChannel.connect(new InetSocketAddress(LOCALHOST_IP, SMTP_PORT));
+            socketChannel.write(ByteBuffer.wrap(message.getBytes()));
+        } catch (IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private boolean messageHasBeenSent(SocketChannel socketChannel, String recipient) throws IOException {
+        ByteBuffer byteBuffer = ByteBuffer.allocateDirect(2048);
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: can't we use something more high-level to consume a string from socketChannel ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 5]
```diff
@@ -49,15 +67,78 @@ public void setup() throws Exception {
...
+                    "content\r\n" +
+                    ".\r\n";
+            socketChannel.connect(new InetSocketAddress(LOCALHOST_IP, SMTP_PORT));
+            socketChannel.write(ByteBuffer.wrap(message.getBytes()));
+        } catch (IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private boolean messageHasBeenSent(SocketChannel socketChannel, String recipient) throws IOException {
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: you don't use the recipient, right ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 6]
```diff
@@ -49,15 +67,78 @@ public void setup() throws Exception {
...
+        return false;
+    }
+
+    private boolean userReceivedMessage(String user) {
+        try {
+            IMAPClient imapClient = new IMAPClient();
+            imapClient.connect(LOCALHOST_IP, IMAP_PORT);
+            imapClient.login(user, PASSWORD);
+            imapClient.select("INBOX");
+            imapClient.fetch("1:1", "ALL");
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: you should close the client

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java (hunk 1 above)
Comment (aduprat) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: I know, but can't find a better way with the java mail API...

File: mailet/base/src/main/java/org/apache/mailet/base/MailetPipelineLogging.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.mailet.base;
+
+import org.slf4j.Logger;
+
+public class MailetPipelineLogging {
+
+    public static void logIn(Logger logger, String mailetName) {
```
Comment (rouazana) [inline mailet/base/src/main/java/org/apache/mailet/base/MailetPipelineLogging.java]: I find logIn/logOut a little confusing (look like a user/password thing). Have you a better name?

File: server/mailet/integration-testing/src/test/resources/logback.xml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        </contextListener>
+
+        <appender name="CONSOLE" class="ch.qos.logback.core.ConsoleAppender">
+                <encoder>
+                        <pattern>%d{HH:mm:ss.SSS} [%-5level] %logger{15} - %msg%n%rEx</pattern>
+                        <immediateFlush>false</immediateFlush>
+                </encoder>
+        </appender>
+
+        <root level="ERROR">
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/resources/logback.xml]: maybe warn?

Comment (aduprat) [inline server/mailet/integration-testing/src/test/resources/logback.xml]: We should only have ERRORS on test, as:
- we will never look at them when building with Jenkins
- when coding a new test, only errors are important (showing our errors)
- they are producing too much noice for development phase.

Comment (rouazana) [inline server/mailet/integration-testing/src/test/resources/logback.xml]: I think we already discussed about this, and I'm still in favor of WARN.
If they are too much WARN, we should either silent them by making it INFO or silent them in configuration else.
WARN are often very useful and can indicate the real cause of an error seen later.

Comment (aduprat) [inline server/mailet/integration-testing/src/test/resources/logback.xml]: Ok

**Comment by mbaechler** [comment]
sorry but it's very hard to review because of the commits you depends on. Please try to solve this issue.

**Comment by aduprat** [comment]
Each comments are handled in #426

**Pull request closed without merging**